webhook.go

```diff
@@ -0,0 +1,210 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package restapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+
+	"sync"
+
+	"github.com/redhat-cne/sdk-go/pkg/util/wait"
+
+	"github.com/gorilla/mux"
+	"github.com/redhat-cne/sdk-go/pkg/pubsub"
+	"github.com/redhat-cne/sdk-go/pkg/types"
+
+	"net/http"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var singleton sync.Once
+
+// WebhookInstance ... is singleton instance
+var WebhookInstance *Webhook
+
+// Webhook defines rest routes server object
+type Webhook struct {
+	port       int
+	apiPath    string
+	closeCh    <-chan struct{}
+	HTTPClient *http.Client
+	httpServer *http.Server
+	status     serverStatus
+	targetPub  *pubsub.PubSub
+}
+
+// HwEvent ... temporary format. TODO: define schema for redfish hw event data
+type HwEvent struct {
+	ID   string `json:"Id"`
+	Name string `json:"Name"`
+}
+
+// InitWebhook is used to supply configurations for rest routes server
+func InitWebhook(port int, apiPath string, targetPub *pubsub.PubSub) *Webhook {
+	singleton.Do(func() {
+		WebhookInstance = &Webhook{
+			port:    port,
+			apiPath: apiPath,
+			status:  notReady,
+			HTTPClient: &http.Client{
+				Transport: &http.Transport{
+					MaxIdleConnsPerHost: 20,
+				},
+				Timeout: 10 * time.Second,
+			},
+			targetPub: targetPub,
+		}
+	})
+	// singleton
+	return WebhookInstance
+}
+
+// EndPointHealthChk checks for rest service health
+func (s *Webhook) EndPointHealthChk() (err error) {
+	log.Info("checking for rest service health\n")
+	for i := 0; i <= 5; i++ {
+		if !s.Ready() {
+			time.Sleep(healthCheckPause)
+			log.Printf("server status %t", s.Ready())
+			continue
+		}
+
+		log.Debugf("health check %s%s ", s.GetHostPath(), "health")
+		response, errResp := http.Get(fmt.Sprintf("%s%s", s.GetHostPath(), "health"))
+		if errResp != nil {
+			log.Errorf("try %d, return health check of the rest service for error  %v", i, errResp)
+			time.Sleep(healthCheckPause)
+			err = errResp
+			continue
+		}
+		if response != nil && response.StatusCode == http.StatusOK {
+			response.Body.Close()
+			log.Infof("rest service returned healthy status")
+			time.Sleep(healthCheckPause)
+			err = nil
+			return
+		}
+		response.Body.Close()
+	}
+	if err != nil {
+		err = fmt.Errorf("error connecting to rest api %s", err.Error())
+	}
+	return
+}
+
+// Port port id
+func (s *Webhook) Port() int {
+	return s.port
+}
+
+//Ready gives the status of the server
+func (s *Webhook) Ready() bool {
+	return s.status == started
+}
+
+// GetHostPath  returns hostpath
+func (s *Webhook) GetHostPath() *types.URI {
+	return types.ParseURI(fmt.Sprintf("http://localhost:%d%s", s.port, s.apiPath))
+}
+
+// Start will start res routes service
+func (s *Webhook) Start() {
+	if s.status == started || s.status == starting {
```

Comment:
status check  needed for default rest API . lets keep it simple for single webhook

Reply:
or you can use directly the pubsub API without using rest service